src/lib.rs

```diff
@@ -875,6 +875,29 @@ where
   }
 }
 
+/// Compute the circuit digest of a [StepCircuit].
+pub fn circuit_digest<
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  C: StepCircuit<G1::Scalar>,
+>(
+  circuit: &C,
+  is_primary_circuit: bool,
+) -> G1::Scalar {
+  let augmented_circuit_params =
+    NovaAugmentedCircuitParams::new(BN_LIMB_WIDTH, BN_N_LIMBS, is_primary_circuit);
+
+  // ro_consts_circuit are parameterized by G2 because the type alias uses G2::Base = G1::Scalar
+  let ro_consts_circuit: ROConstantsCircuit<G2> = ROConstantsCircuit::<G2>::default();
+
+  // Initialize ck for the primary
+  let augmented_circuit: NovaAugmentedCircuit<'_, G2, C> =
+    NovaAugmentedCircuit::new(&augmented_circuit_params, None, circuit, ro_consts_circuit);
+  let mut cs: ShapeCS<G1> = ShapeCS::new();
+  let _ = augmented_circuit.synthesize(&mut cs);
+  cs.r1cs_shape().digest()
```

Comment:
The underlying issue is that we must know the circuit shape in some form to create a valid cache key. Thus, performing `synthesize` is unavoidable (unless there is another way to get the concrete representation of a circuit).

However, as you say, this is expensive. @porcuquine and I discussed this issue at length in [this zulip thread](https://zulip.lurk-lab.com/#narrow/stream/47-systems/topic/Public.20Parameters) and our conclusion was to rely on an "ad-hoc" representation of `(circuit(rc=2), actual_rc)`, which is much faster. So we know by design that this call *should* never be expensive because it is only used [here](https://github.com/lurk-lab/lurk-rs/pull/648/files#diff-cefb89359ae9c92029a66a179ecbcd6f8dc74f04d4d0d77fef25c77f536744c6R598-R608) in lurk-rs. This is hackish but it seems to be the best solution for now.

Reply:
Please add a doc comment to this function and its supernova alter ego to state that this call is quite expensive, then.